core/src/main/java/io/grpc/Metadata.java

```diff
@@ -249,18 +336,24 @@ public boolean containsKey(Key<?> key) {
   public <T> boolean remove(Key<T> key, T value) {
     Preconditions.checkNotNull(key, "key");
     Preconditions.checkNotNull(value, "value");
-    List<MetadataEntry> values = store.get(key.name());
-    if (values == null) {
-      return false;
-    }
-    for (int i = 0; i < values.size(); i++) {
-      MetadataEntry entry = values.get(i);
-      if (!value.equals(entry.getParsed(key))) {
-        continue;
+    for (int i = 0; i < size; i++) {
+      if (Arrays.equals(key.asciiName(), name(i))) {
+        @SuppressWarnings("unchecked")
+        T stored = typedValues[i] != null ? (T) typedValues[i] : key.parseBytes(value(i));
+        if (!value.equals(stored)) {
+          continue;
+        }
+        for (int k = i; k < size - 1; k++) {
```

Comment:
This would be more clear as a `System.arraycopy` (which handles overlapped copies).